linter/src/spawning/spawning.cpp

```diff
@@ -1,35 +1,108 @@
+        throw std::runtime_error("File at NUTC_SPAWNER_BINARY_PATH does not exist");
 
-    if (pid == 0) {
-        std::vector<std::string> args = {
-            "NUTC-linter-spawner", "--uid", uid, "--algoid", sanitized_algoid
-        };
+    return spawner_binary_path;
+}
 
-        std::vector<char*> c_args;
-        for (auto& arg : args)
-            c_args.push_back(arg.data());
-        c_args.push_back(nullptr);
+nutc::lint::lint_result
+LintProcessManager::spawn_client(const std::string& algo_code)
+{
+    static const std::string path{spawner_binary_path()};
+    auto in_pipe = std::make_shared<bp::async_pipe>(io_service_);
+    bp::opstream out_pipe;
 
-        execvp(c_args[0], c_args.data());
+    auto child = std::make_shared<bp::child>(
+        bp::exe(path), bp::std_in<out_pipe, bp::std_out> * in_pipe, io_service_
+    );
 
-        log_e(linting, "Failed to lint algoid {} for uid {}", sanitized_algoid, uid);
+    out_pipe << algo_code << std::flush;
+    out_pipe.pipe().close();
 
-        exit(1);
-    }
-    else if (pid < 0) {
-        log_e(linting, "Failed to fork");
-    }
+    auto kill_timer = std::make_shared<ba::steady_timer>(io_service_);
+    kill_timer->expires_after(ba::chrono::seconds(LINT_AUTO_TIMEOUT_SECONDS));
+
+    nutc::lint::lint_result res;
+
+    kill_timer->async_wait([child, in_pipe, &res](const auto& ec) {
+        if (!ec) {
+            in_pipe->close();
+            res.success = false;
+            res.message += fmt::format(
+                "[linter] FAILED to lint algo\n\nYour code did not execute within {} "
+                "seconds. Check all "
+                "functions to see if you have an infinite loop or infinite "
+                "recursion.\n\nIf you continue to experience this error, "
+                "reach out to #nuft-support.\n",
+                LINT_AUTO_TIMEOUT_SECONDS
+            );
+
+            if (child->running()) {
+                child->terminate();
+            }
+        }
+    });
+
+    auto buffer = std::make_shared<ba::streambuf>();
+    async_read_until(
+        *in_pipe,
+        *buffer,
+        '\n',
+        [buffer, kill_timer, child, in_pipe, &res](const auto& ec, auto) {
+            if (!ec) {
+                kill_timer->cancel();
+
+                std::istream stream(buffer.get());
+                std::string line;
+                std::string message;
+
+                while (std::getline(stream, line)) {
+                    message += line;
+                }
+                if (child->running()) {
+                    child->terminate();
+                }
+
+                auto error = glz::read_json<nutc::lint::lint_result>(res, message);
+                if (error) {
+                    res = {
+                        false,
+                        "Internal server error. Reach out to nuft@u.northwesten.edu "
+                        "for support"
+                    };
+                };
```

Comment:
I'm gonna predict we're gonna run into issues somewhere with this down the line (I ran into a million when doing it for the exchange, but they're more immediately visible there). Just something to keep in mind, not actionable now

Reply:
With telling people to email nutf@u.northwestern.edu?